### PR TITLE
Hoist CI components into reusable actions and workflows

### PR DESCRIPTION
**Pull request by hanno-becker**

### CODE REVIEW

File: .github/actions/lint/action.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+              - $(uname -a)
+              - $(nix --version)
+              - $(astyle --version)
+              - $(${{ matrix.target.cross-prefix }}gcc --version | grep -m1 "")
+              - $(bash --version | grep -m1 "")
+            EOF
+      - name: Set shell
+        shell: bash
+        run: echo SHELL="${{ inputs.use-nix && 'nix develop .#ci-linter -c bash -e {0}' || inputs.custom_shell }}" >> $GITHUB_ENV
+      - name: Run CBMC proofs
```
Comment (potsrevennil) [inline .github/actions/lint/action.yml:37]: Typo

Comment (hanno-becker) [reply .github/actions/lint/action.yml]: Fixed

File: .github/actions/functest/action.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+              - $(bash --version | grep -m1 "")
+            EOF
+      - name: Set shell
+        shell: bash
+        run: echo SHELL="${{ inputs.use-nix && 'nix develop .#ci -c bash -e {0}' || inputs.custom_shell }}" >> $GITHUB_ENV
+      - name: Run functional tests
+        id: func_test
+        shell: ${{ env.SHELL }}
+        run: |
+          tests func --cross-prefix=${{ inputs.cross-prefix }} --cflags ${{ inputs.cflags }} -v
```
Comment (potsrevennil) [inline .github/actions/functest/action.yml:42]: I would like to add this to group the log
```
echo "::group::func_test"
tests func ...
echo "::endgroup::"
```

File: .github/actions/functest/action.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      - name: Run KAT tests
+        if: |
+          success()
+          || steps.func_test.conclusion == 'failure'
+        id: kat_test
+        shell: ${{ env.SHELL }}
+        run: |
+          tests kat --cross-prefix=${{ inputs.cross-prefix }} --cflags ${{ inputs.cflags }} -v
```
Comment (potsrevennil) [inline .github/actions/functest/action.yml:50]: here as well

File: .github/actions/functest/action.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      - name: Run Nistkat tests
+        id: nistkat_test
+        if: |
+          success()
+          || steps.func_test.conclusion == 'failure'
+          || steps.kat_test.conclusion == 'failure'
+        shell: ${{ env.SHELL }}
+        run: |
+          tests nistkat --cross-prefix=${{ inputs.cross-prefix }} --cflags ${{ inputs.cflags }} -v
```
Comment (potsrevennil) [inline .github/actions/functest/action.yml:59]: here as well

File: .github/actions/cbmc/action.yml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      - name: Set shell
+        shell: bash
+        run: echo SHELL="${{ inputs.use-nix && 'nix develop .#ci-cbmc -c bash -e {0}' || inputs.custom_shell }}" >> $GITHUB_ENV
+      - name: Run CBMC proofs
+        shell: ${{ env.SHELL }}
+        run: |
+          cd cbmc/proofs;
+          KYBER_K=2 ./run-cbmc-proofs.py --summarize;
+          KYBER_K=3 ./run-cbmc-proofs.py --summarize;
+          KYBER_K=4 ./run-cbmc-proofs.py --summarize;
```
Comment (potsrevennil) [inline .github/actions/cbmc/action.yml:45]: and maybe here

File: .github/actions/bench/action.yml [hunk 6]
```diff
@@ -32,24 +32,47 @@ inputs:
...
+      with:
+        devShell: ci
+        script: |
+          ARCH=$(uname -m)
+          cat >> $GITHUB_STEP_SUMMARY <<-EOF
+            ## Setup
+            Architecture: $ARCH
+            - $(uname -a)
+            - $(nix --version)
+            - $(astyle --version)
```
Comment (potsrevennil) [inline .github/actions/bench/action.yml:56]: remove this as astyle is not in ci shell

File: .github/actions/functest/action.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        with:
+          devShell: ci
+          script: |
+            ARCH=$(uname -m)
+            cat >> $GITHUB_STEP_SUMMARY <<-EOF
+              ## Setup
+              Architecture: $ARCH
+              - $(uname -a)
+              - $(nix --version)
+              - $(astyle --version)
```
Comment (potsrevennil) [inline .github/actions/functest/action.yml:31]: remove this

Comment (hanno-becker) [reply .github/actions/functest/action.yml]: done

File: .github/actions/bench/action.yml (hunk 6 above)
Comment (hanno-becker) [reply .github/actions/bench/action.yml]: done

File: .github/actions/cbmc/action.yml (hunk 5 above)
Comment (hanno-becker) [reply .github/actions/cbmc/action.yml]: done

File: .github/actions/functest/action.yml (hunk 4 above)
Comment (hanno-becker) [reply .github/actions/functest/action.yml:63]: done

File: .github/actions/functest/action.yml (hunk 3 above)
Comment (hanno-becker) [reply .github/actions/functest/action.yml:52]: done

File: .github/actions/functest/action.yml (hunk 2 above)
Comment (hanno-becker) [reply .github/actions/functest/action.yml:42]: done

**Review by potsrevennil — APPROVED** [review]
All looks good now! After reducing all the boilerplat, I'm happy with the structure now. Thanks for the work!

**Pull request merged**